internal/opa_inspect/opa_inspect.go

```diff
@@ -0,0 +1,151 @@
+// Copyright 2022 Red Hat, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+// Functions for inspecting rego code and annotations
+package opa_inspect
+
+import (
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/open-policy-agent/opa/ast"
+)
+
+func inspectSingle(path, module string) ([]*ast.AnnotationsRef, error) {
+	mod, err := ast.ParseModuleWithOpts(path, module, ast.ParserOptions{ProcessAnnotation: true})
+	if err != nil {
+		return nil, err
+	}
+
+	as, x := ast.BuildAnnotationSet([]*ast.Module{mod})
+	if len(x) > 0 {
+		return nil, err
+	}
+
+	results := make([]*ast.AnnotationsRef, 0, len(mod.Rules))
+	for _, rule := range mod.Rules {
+		results = append(results, as.Chain(rule)...)
+	}
+
+	return results, nil
+}
+
+func inspectMultiple(paths, modules []string) ([]*ast.AnnotationsRef, error) {
+	noPaths := len(paths)
+	if noPaths != len(modules) {
+		return nil, fmt.Errorf("given uneven number of paths and modules: %d != %d", noPaths, len(modules))
+	}
+
+	results := make([]*ast.AnnotationsRef, 0, noPaths)
+	for i := 0; i < noPaths; i++ {
+		r, err := inspectSingle(paths[i], modules[i])
+		if err != nil {
+			return nil, err
+		}
+
+		results = append(results, r...)
+	}
+
+	return results, nil
+}
+
+func isWarnOrDeny(rule *ast.AnnotationsRef) bool {
+	ruleName := rule.Path[len(rule.Path)-1].Value.String()
+	// Fixme: No idea how to compare the ast.Value to a string in a sane
+	// way here. https://pkg.go.dev/github.com/open-policy-agent/opa/ast#Value
```

Comment:
Looking at the docs, does this work? Note lots of pointers in this:

```suggestion
	ruleName := rule.GetRule().Head.Name
```